### PR TITLE
Enable SGC Active Learning

### PR DESCRIPTION
**Pull request by erinaldidb**
# Enable SGC Active Learning

**Branch:** `feature/training_sgc` → `main` • **Commits:** 15 • **Files changed:** 10 • **+1,796 / −12**

## Goal

Bring the full **MONAILabel active-learning loop** — interactive annotation in
OHIF, on-cluster training, MLflow tracking, and checkpoint promotion — into
Pixels and run it directly on a Databricks **Single-GPU Cluster (SGC)**.
Until now the OHIF viewer (deployed as a Databricks App) could only consume
Vista3D inference from the MONAI model-serving endpoint; there was no path for
a clinician to *correct* a segmentation, push the corrected mask back, and
re-train. This PR adds that path end-to-end.

It also lands a handful of DICOMweb gateway fixes that are required to make the
loop work with strict DICOMweb clients (MedDream, the OHIF segmentation
plug-in) and to stream large series metadata efficiently.

---

### 1. New `notebooks/SGC-Training.ipynb` — active-learning workstation

A single notebook that turns a Databricks GPU cluster into a complete
MONAILabel + OHIF workstation:

- **OHIF Viewer + DICOMweb proxy** — a FastAPI app served by Hypercorn on the
  cluster's driver-proxy port. Reverse-proxies:
  - `/dicomweb/*` → the `pixels-dicomweb-gateway` App (with token-exchange
    OBO so the user's identity flows through to UC).
  - `/sqlwarehouse/*` → the workspace SQL warehouse REST API.
  - `/info`, `/infer`, `/train`, `/activelearning`, `/scoring`, `/datastore`,
    `/logs`, `/ohif/*` → the local MONAILabel server on `127.0.0.1`.
  - Injects the `pixels_table` cookie on every static-asset response so the
    gateway knows which Unity Catalog table to query.

- **MONAILabel server (subprocess)** — radiology app launched with
  configurable `models`, `labels`, and `use_pretrained` flags. Runs in its own
  process group (`start_new_session=True`) so notebook cancellations don't
  propagate. The app is copied from the UC Volume to `/tmp` first, because
  FUSE doesn't support `seek()` and that breaks `RotatingFileHandler`,
  TensorBoard event files, and checkpoint append-writes. A custom
  `--log_config` replaces MONAILabel's default `RotatingFileHandler` with a
  plain `FileHandler` for the same reason.

- **Train-proxy middleware** — `MonaiProxyMiddleware` (a Starlette
  `BaseHTTPMiddleware`) intercepts `POST /train/{model}` from the OHIF UI,
  merges the request body with `MONAI_TRAIN_DEFAULTS` (epochs, batch size,
  data-pipeline mode, MLflow tracking config…), and forwards to the local
  MONAILabel server. The defaults are exposed as a single Python dict so
  hyper-parameters not surfaced by the OHIF UI can still be tuned.

- **Lifecycle**: `start_servers()` / `stop_servers()` with graceful Hypercorn
  shutdown via `shutdown_trigger`, `SIGTERM` for the MONAILabel subprocess, and
  a `lsof -ti :PORT` fallback for orphaned listeners.

- **Three operating modes** documented at the top of the notebook:
  | Workflow | Cells |
  | --- | --- |
  | Interactive annotation (pre-trained model) | 2 – 12 |
  | Interactive annotation (custom labels) | 2 – 10, 13 – 14 |
  | Autonomous training (no UI) | 2 – 10, 15 – 17 |

- **Py4J safety** — the MONAILabel subprocess strips `PYSPARK_*` and `PY4J_*`
  env vars so it never attaches to the notebook's JVM gateway. OHIF runs in a
  daemon thread (not a forked child) for the same reason — preventing a
  training subprocess `atexit` handler from corrupting the notebook's gateway.

### 2. New `src/dbx/pixels/modeltraining/` package — programmatic training

Reusable Python interface to the same training pipeline, callable from
notebooks, jobs, or the OHIF train-proxy:

- **`MonaiTrainer`** — initialise once with `app_dir`, `studies_url`,
  `experiment_name`, `pixels_table`, and `train_configs`; then call
  `trainer.train(model=..., labels=..., max_epochs=..., **overrides)`
  repeatedly. Loads the MONAILabel app **in-process** (no HTTP server hop),
  dispatches training on the attached GPU(s), and ships everything to **one
  MLflow run**:
  - Training metrics (via MONAILabel's `MLFlowHandler`).
  - System metrics (GPU/CPU/RAM) via a background `SystemMetricsMonitor` that
    auto-discovers the run created by MONAILabel and attaches to it.
  - The final checkpoint (`checkpoint_final.pt`) as an MLflow artifact.

- **`log_model_weights(...)`** — standalone helper to ship a checkpoint to
  the most recent (or a specified) MLflow run. Used after OHIF-driven training
  to retroactively log the artifact.

- Auto-builds an absolute Databricks workspace experiment path
  (`/Users/<user>/training/monai_training_<safe_table>`) so MLflow doesn't
  reject relative names.

### 3. DICOMweb gateway — study-level WADO-RS metadata

`apps/dicom-web-gateway/`:

- **New endpoint** `GET /api/dicomweb/studies/{study}/metadata` — streams
  DICOM-JSON metadata for **every instance in a study**, mirroring the
  existing series-level variant. Required by OHIF's "load full study
  metadata up-front" mode and by some segmentation plug-ins.
  - `utils/queries.py` — new `build_study_metadata_query()` (single
    `SELECT meta FROM <pixels_table> WHERE 0020000D = :study_uid`).
  - `utils/wrapper.py` — new `retrieve_study_metadata()` reusing the same
    streaming Arrow-batch pattern as `retrieve_series_metadata()` (memory
    proportional to one batch, not the full result set).
  - `utils/handlers/_wado.py` — `dicomweb_wado_study_metadata()` returning a
    `StreamingResponse` with `application/dicom+json`.
  - Wired up in `app.py` and listed in the `/api/dicomweb` root index.

### 4. DICOMweb gateway — strict-parser metadata sanitization

The `meta` VARIANT column carries pipeline-injected fields (`file_size`,
`hash`, `has_pixel`, …) that aren't valid DICOMweb attributes. Strict viewers
(MedDream's `AttributeElement` parser, the OHIF segmentation plug-in) reject
the whole response when they encounter them.

- **New `utils/dicom_tags.py::filter_dicom_json()`** — keeps only keys
  matching the canonical 8-hex-character DICOM tag ID format
  (`^[0-9A-Fa-f]{8}$`). Applied to both the per-instance series metadata
  stream and the new per-study metadata stream.

### 5. DICOMweb gateway — tunable stream chunk size

- **New env var `PIXELS_STREAM_CHUNK_SIZE`** (default 1 MiB, was hard-coded
  256 KiB) propagated through `BufferedStreamReader`, `stream_file()`,
  `ProgressiveFileStreamer`, and `_compute_bot_via_stream()`. Improves
  throughput on large multi-frame instances without raising memory pressure.

### 6. Minor fixes

- `apps/dicom-web-gateway/utils/handlers/_stow.py` — `table_scoped_base` and
  `dest_path` are now constructed **after** the pixels-table is resolved.
  Previously they were referencing the global default before the
  per-request override was applied, producing wrong destination paths under
  multi-table deployments.

---

## Files changed

```
 apps/dicom-web-gateway/app.py                     |   10 +
 apps/dicom-web-gateway/utils/dicom_io.py          |   14 +-
 apps/dicom-web-gateway/utils/dicom_tags.py        |   24 +
 apps/dicom-web-gateway/utils/handlers/__init__.py |    2 +
 apps/dicom-web-gateway/utils/handlers/_wado.py    |   15 +
 apps/dicom-web-gateway/utils/queries.py           |   24 +
 apps/dicom-web-gateway/utils/wrapper.py           |   50 +-
 notebooks/SGC-Training.ipynb                      | 1333 +++++++++
 src/dbx/pixels/modeltraining/__init__.py          |   26 +
 src/dbx/pixels/modeltraining/_training.py         |  310 +++
 10 files changed, 1796 insertions(+), 12 deletions(-)
```

## How to use (TL;DR)

1. Spin up an SGC (e.g. A10 / A100), `%pip install` the MONAILabel wheel + the
   `databricks-pixels` wheel from the UC Volume populated by `deploy_prep`.
2. Run `notebooks/SGC-Training.ipynb` cells 2 – 10.
3. **For annotation + UI-driven training** — run `start_servers(...)`, open
   the printed OHIF URL, segment a study, click *Train* in the OHIF panel.
   Metrics + checkpoint land in
   `/Users/<you>/training/monai_training_<table>` in MLflow.
4. **For autonomous training (e.g. from a job)** — instantiate `MonaiTrainer`
   and call `trainer.train(model="segmentation", labels="lung_left,lung_right",
   max_epochs=100)`. Run-id is returned in `result["mlflow_run_id"]`.

## Validation

- OHIF viewer loads, `GET /api/dicomweb/studies/{uid}/metadata` returns
  filtered DICOM-JSON (verified end-to-end against MedDream's strict parser).
- Series-metadata stream still parses in OHIF after the `filter_dicom_json`
  hook is applied.
- `MonaiTrainer.train(...)` produces an MLflow run with training metrics,
  system metrics, and `checkpoint_final.pt` as an artifact — all under one
  `run_id`.
- OHIF *Train* button in MONAI-Label mode produces a run identical in shape
  to the autonomous path.
- STOW-RS still routes to the correct `dest_path` after the
  `table_scoped_base` reordering.

### CODE REVIEW

File: notebooks/training/AIRuntime-ActiveLearning.ipynb
Comment (dmoore247) [inline notebooks/training/AIRuntime-ActiveLearning.ipynb:1]: Please put notebook into `./notebooks/training/SGC-ActiveLearningTraining.ipynb`
The exact notebook name isn't so important to me, but let's plan for multiple training notebooks.

File: apps/dicom-web-gateway/utils/queries.py [hunk 1]
```diff
@@ -284,6 +284,30 @@ def build_series_metadata_query(
...
+    Build a WADO-RS query to retrieve raw DICOM JSON metadata for every
+    instance in a study.
+
+    Returns only the ``meta`` VARIANT column — the caller streams it
+    directly as JSON without deserialization.
+    """
+    validate_table_name(pixels_table)
+    query = """
+    SELECT meta
+    FROM IDENTIFIER(%(pixels_table)s)
```
Comment (dmoore247) [inline apps/dicom-web-gateway/utils/queries.py:329]: @erinaldidb 
I see IDENTIFIER and sql_params are being used. This is a good sign.
Could you run a security assessment code review prompt over this file to see if more is needed?
Was there any additional sanitization performed on these two parameters upstream?

**Review by dmoore247 — CHANGES_REQUESTED** [review]
Looks good, a couple items to look at.

File: notebooks/training/AIRuntime-ActiveLearning.ipynb
Comment (dmoore247) [reply notebooks/training/AIRuntime-ActiveLearning.ipynb:1]: SGC has effectively been renamed to AI Runtime

File: apps/dicom-web-gateway/utils/queries.py (hunk 1 above)
Comment (erinaldidb) [reply apps/dicom-web-gateway/utils/queries.py:329]: Few changes applied, implemented limit and offset max values

**Review by dmoore247 — APPROVED** [review]
LGTM!
